subsys/net/lib/latmon/latmon.c

```diff
@@ -0,0 +1,568 @@
+
+struct latmon_net_request {
+	uint32_t period_usecs;
+	uint32_t histogram_cells;
+} __packed;
+
+struct latmon_net_data {
+	int32_t sum_lat_hi;
+	int32_t sum_lat_lo;
+	int32_t min_lat;
+	int32_t max_lat;
+	uint32_t overruns;
+	uint32_t samples;
+} __packed;
+
+/* Private IPC: Zephyr application to Latmon service */
+struct latmon_message {
+	net_latmon_measure_t measure_func;
+	int latmus; /* latmus connection */
+};
+
+K_MSGQ_DEFINE(latmon_msgq, sizeof(struct latmon_message), 2, 4);
+
+/*
+ * Note: Using a small period (e.g., less than 100 microseconds) may result in
+ * the reporting too good interrupt latencies during a short test due to cache
+ * effects.
+ */
+struct latmus_conf {
+	uint32_t max_samples;
+	uint32_t period; /* in usecs */
+	uint32_t cells;
+};
+
+/*
+ * Each cell represents a 1 usec timespan.
+ * note: the sampling period cannot be longer than 1 sec.
+ */
+#define MAX_SAMPLING_PERIOD_USEC 1000000
+#define HISTOGRAM_CELLS_MAX 1000
+struct latmon_data {
+	bool warmed; /* sample data can be used */
+	uint32_t histogram[HISTOGRAM_CELLS_MAX];
+	uint32_t current_samples;
+	uint32_t overruns;
+	uint32_t min_lat;
+	uint32_t max_lat;
+	uint64_t sum_lat;
+};
+
+/* Message queue for sample data transfers */
+K_MSGQ_DEFINE(xfer_msgq, sizeof(struct latmon_data), 10, 4);
+
+/* Network transfer thread: sends data to Latmus  */
+#define XFER_THREAD_STACK_SIZE CONFIG_NET_LATMON_XFER_THREAD_STACK_SIZE
+#define XFER_THREAD_PRIORITY CONFIG_NET_LATMON_XFER_THREAD_PRIORITY
+K_THREAD_STACK_DEFINE(xfer_thread_stack, XFER_THREAD_STACK_SIZE);
+static struct k_thread xfer_thread;
+
+/* Latmon thread: receives application requests */
+#define LATMON_THREAD_PRIORITY CONFIG_NET_LATMON_THREAD_PRIORITY
+#define LATMON_STACK_SIZE CONFIG_NET_LATMON_THREAD_STACK_SIZE
+
+/* Monitor thread: performs the sampling */
+#define MONITOR_THREAD_PRIORITY CONFIG_NET_LATMON_MONITOR_THREAD_PRIORITY
+#define MONITOR_STACK_SIZE CONFIG_NET_LATMON_MONITOR_THREAD_STACK_SIZE
+static K_THREAD_STACK_DEFINE(monitor_stack, MONITOR_STACK_SIZE);
+
+static struct k_thread monitor_thread;
+static k_tid_t monitor_tid;
+static bool abort_monitor;
+
+/* Synchronization */
+static K_SEM_DEFINE(latmon_done, 0, 1);
+static K_SEM_DEFINE(monitor_done, 0, 1);
+
+static ssize_t send_net_data(int latmus, const void *buf, size_t count)
+{
+	ssize_t total_written = 0;
+	ssize_t bytes_written;
+
+	while (count > 0) {
+		const char *p = (const char *)buf + total_written;
+
+		bytes_written = zsock_send(latmus, p, count, 0);
+		if (bytes_written < 0) {
+			if (errno == EINTR) {
+				continue;
+			}
+			return -1;
+		}
+		if (bytes_written == 0) {
+			break;
+		}
+
+		total_written += bytes_written;
+		count -= bytes_written;
+	}
+
+	return total_written;
+}
+
+static int send_sample_data(int latmus, struct latmon_data *data)
+{
+	struct latmon_net_data ndata = {
+		.sum_lat_lo = htonl(data->sum_lat & 0xffffffff),
+		.sum_lat_hi = htonl(data->sum_lat >> 32),
```

Comment:
can't `htonll` be used to simplify this?

Reply:
this is the interface with latmus (ie: https://source.denx.de/Xenomai/xenomai4/libevl/-/blob/master/latmus/gpio.c?ref_type=heads#L274 ) . We could change it but then it would affect previous users for not much gain.